### PR TITLE
UCT/ZE/TEST: expand ZE RMA coverage and fix ze_copy rkey_compare path

### PR DESCRIPTION
**Pull request by yafshar**
## What?
This PR expands ZE test coverage in UCT gtests and fixes a ZE MD callback issue
that caused a crash in existing tests.

- Expand ZE RMA test coverage for the ze_copy transport in test_ze.cc
- Extend ZE-related MD test coverage and expectations in test_md.cc
- Register ZE gtest sources in Makefile.am
- Update ZE test helpers in test_p2p_rma.cc and uct_test.h
- Fix ze_copy component rkey compare callback and mem_dereg parameter handling
  in ze_copy_md.c
- Fix ze_copy_md.c aligned with UCX style guidelines

## Why?
ZE paths lacked sufficient automated coverage for ze_copy transport and MD
behavior, and an existing ZE MD test exposed a real failure:

- The ze_copy MD test path crashed during rkey comparison due to missing component callback wiring.
- Additional coverage is required to consistently validate ZE host, device, and managed memory cases across RMA and MD operations.

### CODE REVIEW

File: test/gtest/uct/test_md.cc [hunk 1]
```diff
@@ -1186,6 +1179,110 @@ UCS_TEST_SKIP_COND_P(test_md_non_blocking, reg,
...
 
+#define UCT_MD_ZE_INSTANTIATE_TEST_CASE(_test_case) \
+    _UCT_MD_INSTANTIATE_TEST_CASE(_test_case, ze_cpy)
+
+class test_md_ze : public test_md {
+};
+
+UCS_TEST_SKIP_COND_P(test_md_ze, mem_query_dmabuf,
+                     !check_caps(UCT_MD_FLAG_ALLOC | UCT_MD_FLAG_REG |
+                                 UCT_MD_FLAG_REG_DMABUF))
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc]: can this test be generic, not just for ZE? since it's testing the generic dmabuf interface

Comment (yafshar) [reply test/gtest/uct/test_md.cc]: Thanks, updated to a generic MD test (test_md_dmabuf) and instantiated with UCT_MD_INSTANTIATE_TEST_CASE.

Comment (yafshar) [reply test/gtest/uct/test_md.cc]: @yosefe please let me know if anything else is needed!

**Comment by yafshar** [comment]
Clarification: this PR is not blocked on #11321, but without it the ZE md_buffer test path will be skipped due to missing ZE support in gtest mem_buffer/UCM hooks.

**Comment by yafshar** [comment]
Hi @yosefe, just a gentle follow‑up in case this slipped through.
Happy to address any questions or make changes if needed.
Thanks!

**Comment by yafshar** [comment]
Hi @yosefe @tvegas1  just a gentle follow‑up in case this slipped through.
Happy to address any questions or make changes if needed.
Thanks!

File: src/uct/ze/copy/ze_copy_md.c [hunk 2]
```diff
@@ -10,29 +10,28 @@
...
 #include <ucs/memory/memtype_cache.h>
 
+#include <limits.h>
+#include <string.h>
+
 
-static ucs_config_field_t uct_ze_copy_md_config_table[] = {
-    {"", "", NULL, ucs_offsetof(uct_ze_copy_md_config_t, super),
-     UCS_CONFIG_TYPE_TABLE(uct_md_config_table)},
+static ucs_config_field_t uct_ze_copy_md_config_table[] =
```
Comment (tvegas1) [inline src/uct/ze/copy/ze_copy_md.c]: maybe leave it untouched as it is only whitespace changes?

File: src/uct/ze/copy/ze_copy_md.c [hunk 3]
```diff
@@ -78,21 +77,20 @@ uct_ze_copy_mem_alloc(uct_md_h tl_md, size_t *length_p, void **address_p,
 
     switch (mem_type) {
     case UCS_MEMORY_TYPE_ZE_HOST:
-        status = UCT_ZE_FUNC_LOG_ERR(zeMemAllocHost(md->ze_context, &host_desc,
```
Comment (tvegas1) [inline src/uct/ze/copy/ze_copy_md.c:81]: remove all whitespace changes in both switch case.

File: src/uct/ze/copy/ze_copy_md.c [hunk 4]
```diff
@@ -309,15 +335,17 @@ uct_ze_copy_md_open(uct_component_h component, const char *md_name,
     /* Get sub-device by device_ordinal (global sub-device ID) */
     subdevice = uct_ze_base_get_subdevice_by_global_id(config->device_ordinal);
     if (subdevice == NULL) {
-        ucs_error("Failed to get sub-device at ordinal %d", config->device_ordinal);
```
Comment (tvegas1) [inline src/uct/ze/copy/ze_copy_md.c:312]: remove all unrelated whitespace changes

File: src/uct/ze/copy/ze_copy_md.c [hunk 5]
```diff
@@ -133,6 +131,30 @@ uct_ze_copy_mem_reg(uct_md_h md, void *address, size_t length,
     return UCS_OK;
 }
 
+static ucs_status_t
+uct_ze_copy_mem_dereg(uct_md_h md, const uct_md_mem_dereg_params_t *params)
+{
+    uct_mem_h memh;
+    unsigned flags;
+
+    memh = (params->field_mask & UCT_MD_MEM_DEREG_FIELD_MEMH) ? params->memh :
```
Comment (tvegas1) [inline src/uct/ze/copy/ze_copy_md.c]: replace most of the content of this function by `UCT_MD_MEM_DEREG_CHECK_PARAMS(params, 0)`?

File: src/uct/ze/copy/ze_copy_md.c [hunk 6]
```diff
@@ -145,7 +167,7 @@ static ucs_status_t
 uct_ze_copy_md_query_attributes(uct_md_h md, const void *addr, size_t length,
                                 ucs_memory_info_t *mem_info, int *dmabuf_fd)
 {
-    uct_ze_copy_md_t *ze_md                  = ucs_derived_of(md, uct_ze_copy_md_t);
+    uct_ze_copy_md_t *ze_md = ucs_derived_of(md, uct_ze_copy_md_t);
```
Comment (tvegas1) [inline src/uct/ze/copy/ze_copy_md.c]: remove whitespace changes

File: src/uct/ze/copy/ze_copy_md.c [hunk 7]
```diff
@@ -339,20 +367,26 @@ uct_ze_copy_md_open(uct_component_h component, const char *md_name,
...
+    .cm_open            = (uct_component_cm_open_func_t)
+            ucs_empty_function_return_unsupported,
     .rkey_unpack        = uct_ze_copy_rkey_unpack,
-    .rkey_ptr           = (uct_component_rkey_ptr_func_t)ucs_empty_function_return_unsupported,
-    .rkey_release       = (uct_component_rkey_release_func_t)ucs_empty_function_return_success,
+    .rkey_ptr           = (uct_component_rkey_ptr_func_t)
+            ucs_empty_function_return_unsupported,
+    .rkey_release       = (uct_component_rkey_release_func_t)
+            ucs_empty_function_return_success,
+    .rkey_compare       = uct_base_rkey_compare,
```
Comment (tvegas1) [inline src/uct/ze/copy/ze_copy_md.c:353]: remove all other whitespaces changes around this one

File: test/gtest/uct/test_md.cc [hunk 8]
```diff
@@ -33,22 +33,9 @@ extern "C" {
...
-                   cuda_ipc, \
-                   rocm_cpy, \
-                   rocm_ipc, \
-                   ib, \
-                   ugni, \
-                   gdr_copy, \
-                   sockcm, \
-                   rdmacm \
-                   )
+    UCS_PP_FOREACH(_UCT_MD_INSTANTIATE_TEST_CASE, _test_case, knem, cma, \
```
Comment (tvegas1) [inline test/gtest/uct/test_md.cc]: remove whitespace changes

File: test/gtest/uct/test_md.cc [hunk 9]
```diff
@@ -1190,6 +1183,108 @@ UCS_TEST_SKIP_COND_P(test_md_non_blocking, reg,
...
+    }
+}
+
+UCS_TEST_P(test_md_dmabuf, mem_query_dmabuf_offset)
+{
+    if ((md_attr().dmabuf_mem_types & md_attr().access_mem_types) == 0) {
+        UCS_TEST_SKIP_R("MD does not expose dmabuf memory query");
+    }
+
+    const size_t size   = 4096;
```
Comment (tvegas1) [inline test/gtest/uct/test_md.cc]: do we need to use page size instead of 4kB here?

File: test/gtest/uct/test_md.cc [hunk 10]
```diff
@@ -1190,6 +1183,108 @@ UCS_TEST_SKIP_COND_P(test_md_non_blocking, reg,
...
+class test_md_dmabuf : public test_md {
+};
+
+UCS_TEST_P(test_md_dmabuf, mem_query_dmabuf)
+{
+    if ((md_attr().dmabuf_mem_types & md_attr().access_mem_types) == 0) {
+        UCS_TEST_SKIP_R("MD does not expose dmabuf memory query");
+    }
+
+    const size_t size = 4096;
```
Comment (tvegas1) [inline test/gtest/uct/test_md.cc]: maybe use page size here

File: test/gtest/uct/ze/test_ze.cc [hunk 11]
```diff
@@ -0,0 +1,507 @@
...
+                    ucs_max((size_t)1,
+                            (size_t)sender().iface_attr().cap.put.min_zcopy));
+
+    test_xfer(static_cast<send_func_t>(&uct_p2p_rma_test::put_zcopy), length,
+              TEST_UCT_FLAG_SEND_ZCOPY, UCS_MEMORY_TYPE_ZE_DEVICE);
+}
+
+UCS_TEST_P(test_ze_copy_rma, put_zcopy_ze_device_range)
+{
+    if (!check_caps(UCT_IFACE_FLAG_PUT_ZCOPY) ||
```
Comment (tvegas1) [inline test/gtest/uct/ze/test_ze.cc]: use UCS_TEST_COND_P instead of in-test skipping?

File: test/gtest/uct/ze/test_ze.cc [hunk 12]
```diff
@@ -0,0 +1,507 @@
...
+
+    test_xfer_multi_mem_type(static_cast<send_func_t>(
+                                     &uct_p2p_rma_test::put_zcopy),
+                             min_length, max_length, TEST_UCT_FLAG_SEND_ZCOPY,
+                             UCS_MEMORY_TYPE_ZE_DEVICE);
+}
+
+UCS_TEST_P(test_ze_copy_rma, put_short_ze_device)
+{
+    if (!check_caps(UCT_IFACE_FLAG_PUT_SHORT) ||
```
Comment (tvegas1) [inline test/gtest/uct/ze/test_ze.cc]: UCS_TEST_COND_P

File: test/gtest/uct/ze/test_ze.cc [hunk 13]
```diff
@@ -0,0 +1,507 @@
...
+}
+
+UCS_TEST_P(test_ze_copy_rma, put_zcopy_ze_managed)
+{
+    if (!check_caps(UCT_IFACE_FLAG_PUT_ZCOPY) ||
+        !supports_mem_type(UCS_MEMORY_TYPE_ZE_MANAGED)) {
+        UCS_TEST_SKIP_R("ze_copy PUT zcopy ZE managed memory is not available");
+    }
+
+    size_t length =
```
Comment (tvegas1) [inline test/gtest/uct/ze/test_ze.cc]: use common function in all tests and pass memory type as parameter

File: test/gtest/uct/ze/test_ze.cc [hunk 14]
```diff
@@ -0,0 +1,507 @@
...
+                            (size_t)sender().iface_attr().cap.get.max_short));
+
+    test_xfer(static_cast<send_func_t>(&uct_p2p_rma_test::get_short), length,
+              TEST_UCT_FLAG_RECV_ZCOPY, UCS_MEMORY_TYPE_ZE_HOST);
+}
+
+UCS_TEST_P(test_ze_copy_rma, get_short_ze_host_range)
+{
+    if (!check_caps(UCT_IFACE_FLAG_GET_SHORT) ||
+        !supports_mem_type(UCS_MEMORY_TYPE_ZE_HOST)) {
```
Comment (tvegas1) [inline test/gtest/uct/ze/test_ze.cc]: all tests above might be able to use common function with memory type as parameter (might other param needed to add)

File: src/uct/ze/copy/ze_copy_md.c (hunk 2 above)
Comment (yafshar) [reply src/uct/ze/copy/ze_copy_md.c]: done!

File: src/uct/ze/copy/ze_copy_md.c (hunk 3 above)
Comment (yafshar) [reply src/uct/ze/copy/ze_copy_md.c:81]: done!

File: src/uct/ze/copy/ze_copy_md.c (hunk 4 above)
Comment (yafshar) [reply src/uct/ze/copy/ze_copy_md.c:312]: done!

File: src/uct/ze/copy/ze_copy_md.c (hunk 6 above)
Comment (yafshar) [reply src/uct/ze/copy/ze_copy_md.c]: done!

File: src/uct/ze/copy/ze_copy_md.c (hunk 7 above)
Comment (yafshar) [reply src/uct/ze/copy/ze_copy_md.c:353]: done!

File: test/gtest/uct/test_md.cc (hunk 8 above)
Comment (yafshar) [reply test/gtest/uct/test_md.cc]: done!

File: src/uct/ze/copy/ze_copy_md.c (hunk 5 above)
Comment (yafshar) [reply src/uct/ze/copy/ze_copy_md.c]: done, replaced with `UCT_MD_MEM_DEREG_CHECK_PARAMS(params, 0)`

File: test/gtest/uct/test_md.cc (hunk 9 above)
Comment (yafshar) [reply test/gtest/uct/test_md.cc]: done, using `ucs_get_page_size()`

File: test/gtest/uct/test_md.cc (hunk 10 above)
Comment (yafshar) [reply test/gtest/uct/test_md.cc]: done, using `ucs_get_page_size()`

File: test/gtest/uct/ze/test_ze.cc (hunk 11 above)
Comment (yafshar) [reply test/gtest/uct/ze/test_ze.cc]: done, switched to `UCS_TEST_SKIP_COND_P`

File: test/gtest/uct/ze/test_ze.cc (hunk 12 above)
Comment (yafshar) [reply test/gtest/uct/ze/test_ze.cc]: done, switched to `UCS_TEST_SKIP_COND_P`

File: test/gtest/uct/ze/test_ze.cc (hunk 13 above)
Comment (yafshar) [reply test/gtest/uct/ze/test_ze.cc]: done, consolidated into run_single/run_range helpers parametrized by memory type

File: test/gtest/uct/ze/test_ze.cc (hunk 14 above)
Comment (yafshar) [reply test/gtest/uct/ze/test_ze.cc]: done, consolidated into run_single/run_range helpers parametrized by memory type

**Comment by yafshar** [comment]
@tvegas1 thanks for the review. All comments addressed and validated on Intel GPU (ze_copy): focused ze_cpy MD tests pass, ze_caps_and_mem_types passes, and ze_copy RMA transfer cases are enumerated but skip pending gtest mem_buffer ZE allocation support (#11321).

File: test/gtest/uct/ze/test_ze.cc [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+
+    run_range(static_cast<send_func_t>(&uct_p2p_rma_test::get_short),
+              TEST_UCT_FLAG_RECV_ZCOPY, min_length, max_length);
+}
+
+UCS_TEST_SKIP_COND_P(test_ze_copy_rma, ze_caps_and_mem_types,
+                     !check_caps(UCT_IFACE_FLAG_GET_ZCOPY |
+                                 UCT_IFACE_FLAG_PUT_ZCOPY))
+{
+    EXPECT_TRUE(check_caps(UCT_IFACE_FLAG_GET_ZCOPY));
```
Comment (tvegas1) [inline test/gtest/uct/ze/test_ze.cc]: seems redundant with condition in UCS_TEST_SKIP_COND_P?

File: test/gtest/uct/test_md.cc [hunk 16]
```diff
@@ -1190,6 +1197,111 @@ UCS_TEST_SKIP_COND_P(test_md_non_blocking, reg,
...
+        }
+    }
+
+    if (tested == 0) {
+        UCS_TEST_SKIP_R("MD does not expose dmabuf for any supported memory "
+                        "type");
+    }
+}
+
+UCS_TEST_P(test_md_dmabuf, mem_query_dmabuf_offset)
```
Comment (tvegas1) [inline test/gtest/uct/test_md.cc]: maybe the test above with offset=0 can be removed and we only keep that `mem_query_dmabuf_offset` which also checks offset. could also eventually be renamed `mem_query_dmabuf`.

Comment (yafshar) [reply test/gtest/uct/test_md.cc]: Good call on consolidating. One caveat: `mem_query_dmabuf_offset` only iterates memory types that export dmabuf (it continues otherwise), so it doesn't cover the two branches the `offset=0` test does which are the negative case (memory type without a dmabuf fd → `UCT_DMABUF_FD_INVALID`) and the error case (`UCS_ERR_UNSUPPORTED`/`UCS_ERR_INVALID_ADDR`). So deleting it outright would drop that coverage.

I'll merge the two into a single `mem_query_dmabuf` that loops over offsets `{0, 128}`, keeping the negative/error-path checks.

File: test/gtest/uct/ze/test_ze.cc (hunk 15 above)
Comment (yafshar) [reply test/gtest/uct/ze/test_ze.cc]: done!

File: test/gtest/uct/test_md.cc (hunk 16 above)
Comment (yafshar) [reply test/gtest/uct/test_md.cc]: done! pushed.

**Review by tvegas1 — DISMISSED** [review]

**Comment by yafshar** [comment]
@tvegas1 can we merge this PR?

**Comment by yafshar** [comment]
Hi @yosefe @tvegas1 just a gentle follow‑up. Can we merge this PR?

File: test/gtest/uct/test_md.cc [hunk 17]
```diff
@@ -276,7 +277,10 @@ bool test_md::is_device_detected(ucs_memory_type_t mem_type)
 {
     return (mem_type != UCS_MEMORY_TYPE_ROCM) &&
            (mem_type != UCS_MEMORY_TYPE_ROCM_MANAGED) &&
-           (mem_type != UCS_MEMORY_TYPE_CUDA_MANAGED);
+           (mem_type != UCS_MEMORY_TYPE_CUDA_MANAGED) &&
```
Comment (brminich) [inline test/gtest/uct/test_md.cc:281]: pls add Intel copyrights to this file

**Review by brminich — COMMENTED** [review]
pls update all affected files with Intel copyright

**Comment by yafshar** [comment]
> pls update all affected files with Intel copyright

@brminich thanks — Intel copyright added to all files touched by this PR. 

Also resolved a conflict merging master.

One update on my earlier note: now that #11321 is merged, gtest `mem_buffer`
supports ZE allocation, so the `ze_copy` RMA transfer cases actually execute
instead of skipping. Re-verified on Intel GPU — ZE RMA transfer cases,
`ze_caps_and_mem_types`, and the ze_copy MD tests all pass.

Ready to merge from my side.

**Review by tvegas1 — APPROVED** [review]

**Review by brminich — APPROVED** [review]
